activerecord/test/cases/validations/uniqueness_validation_test.rb

```diff
@@ -314,6 +314,51 @@ def test_validate_case_insensitive_uniqueness_with_special_sql_like_chars
     assert t3.save, "Should save t3 as unique"
   end
 
+  if current_adapter?(:Mysql2Adapter)
+    def test_deprecate_validate_uniqueness_mismatched_collation
+      Topic.validates_uniqueness_of(:author_email_address)
+
+      topic1 = Topic.new(author_email_address: "david@loudthinking.com")
+      topic2 = Topic.new(author_email_address: "David@loudthinking.com")
+
+      assert_equal 1, Topic.where(author_email_address: "david@loudthinking.com").count
+
+      assert_deprecated do
+        assert_not topic1.valid?
+        assert_not topic1.save
+        assert topic2.valid?
+        assert topic2.save
+      end
+
+      assert_equal 2, Topic.where(author_email_address: "david@loudthinking.com").count
+      assert_equal 2, Topic.where(author_email_address: "David@loudthinking.com").count
+    end
+  end
+
+  def test_validate_case_sensitive_uniqueness_by_default
+    Topic.validates_uniqueness_of(:author_email_address)
+
+    topic1 = Topic.new(author_email_address: "david@loudthinking.com")
+    topic2 = Topic.new(author_email_address: "David@loudthinking.com")
+
+    assert_equal 1, Topic.where(author_email_address: "david@loudthinking.com").count
+
+    ActiveSupport::Deprecation.silence do
+      assert_not topic1.valid?
+      assert_not topic1.save
+      assert topic2.valid?
+      assert topic2.save
+    end
+
+    if current_adapter?(:Mysql2Adapter)
+      assert_equal 2, Topic.where(author_email_address: "david@loudthinking.com").count
+      assert_equal 2, Topic.where(author_email_address: "David@loudthinking.com").count
```

Comment:
That happening this (break the uniqueness contract) easily without race condition is what I want to prevent, which frequently experienced to me as MySQL DBA.